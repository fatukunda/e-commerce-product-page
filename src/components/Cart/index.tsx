import React, { FunctionComponent } from "react";

import Button from "../Button";
import deleteIcon from "../../assets/images/trash.svg";
import { ICartItem } from "src/store/slices/cartSlice";
import { removeFromCart } from "src/store/slices/cartSlice";

import { useAppDispatch, useAppSelector } from "src/store/hooks";

const Cart: FunctionComponent = (): JSX.Element => {
  const dispatch = useAppDispatch();
  const cartItems: ICartItem[] = useAppSelector((state) => state.cart.items);
  const itemCount: number = useAppSelector((state) => state.cart.itemCount);
  return (
    <div className="absolute top-[70px] left-1/2 z-10 w-[95%] max-w-md -translate-x-1/2 rounded-lg bg-white shadow-2xl shadow-tertiary/70 md:right-0 md:top-20 md:-translate-x-16 lg:top-28 lg:right-0 lg:w-96 lg:translate-x-56">
      <span className="border-secondary-dark-dark block border-b border-solid p-6 font-bold">
        Cart
      </span>

      <div className={`w-full ${!itemCount ? "py-20" : "p-6"}`}>
        {!itemCount ? (
          <span className="inline-block w-full text-center font-bold text-tertiary-light">
            Your cart is empty
          </span>
        ) : (
          <div>
            {cartItems.map((item, idx) => (
              <div key={idx} className="mb-6 flex items-center gap-4">
                <img
                  src={item.product.image.src}
                  alt={item.product.title}
                  className="w-12 rounded-md"
                />

                <div className="mr-auto flex flex-col font-bold text-tertiary-dark">
                  <span>{item.product.title}</span>
                  <span className="font-normal">
                  <span className="pr-2">{item.variant.title}</span>
                    â‚¬{item.variant.price} x {item.quantity}{" "}
                    <span className="font-bold text-tertiary">
                      ${(Number(item.variant.price) * item.quantity).toFixed(2)}
                    </span>
                  </span>
                </div>

                <button
                  type="button"
                  onClick={() => dispatch(removeFromCart(item))}
                >
                  <img src={deleteIcon} alt="delete" />
                </button>
              </div>
            ))}

            <Button>
              <span>Checkout</span>
            </Button>
          </div>
        )}
      </div>
    </div>
  );
};

export default Cart;
