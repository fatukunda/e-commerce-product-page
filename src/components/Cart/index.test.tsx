import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";

import Cart from ".";
import mockStore from "../mocks/store";

const renderCart = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <Cart />
    </ReduxProvider>
  );
describe("Tests for <Cart/>", () => {
  afterEach(cleanup);
  test("AppHeader: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <Cart />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders without crashing", () => {
    renderCart();
  });
  it('Displays heading "Shopping Cart"', () => {
    const { getByText } = renderCart();
    expect(getByText("Your shopping Cart")).toBeInTheDocument();
  });
  it("Displays product variables correctly", () => {
    const { getByText } = renderCart();
    expect(getByText("nu3 Bio Acerola")).toBeInTheDocument();
    expect(getByText("150 g")).toBeInTheDocument();
    expect(getByText("€18.99 x 1")).toBeInTheDocument();
    expect(getByText("€18.99")).toBeInTheDocument();
  });
  it('Removes item from the cart', () => {
    const { getAllByTestId, getByText } = renderCart();
    const deleteBtns = getAllByTestId('delete-btn');

    fireEvent.click(deleteBtns[0])
    expect(getByText('Your cart is empty')).toBeInTheDocument();
});
});
