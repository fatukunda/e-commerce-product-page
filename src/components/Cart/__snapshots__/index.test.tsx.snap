// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Cart/> AppHeader: Snapshot 1`] = `
<div
  className="absolute top-[70px] left-1/2 z-10 w-[95%] max-w-md -translate-x-1/2 rounded-lg bg-white shadow-2xl shadow-tertiary/70 md:right-0 md:top-20 md:-translate-x-16 lg:top-28 lg:right-0 lg:w-96 lg:translate-x-56"
>
  <span
    className="border-secondary-dark-dark block border-b border-solid p-6 font-bold"
  >
    Your shopping Cart
  </span>
  <div
    className="w-full p-6"
  >
    <div>
      <div
        className="mb-6 flex items-center gap-4"
      >
        <img
          alt="nu3 Bio Acerola"
          className="w-12 rounded-md"
          src="https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
        />
        <div
          className="mr-auto flex flex-col font-bold text-tertiary-dark"
        >
          <span>
            nu3 Bio Acerola
          </span>
          <span
            className="font-normal"
          >
            <span
              className="pr-2"
            >
              150 g
            </span>
            €
            18.99
             x 
            1
             
            <span
              className="font-bold text-tertiary"
            >
              €
              18.99
            </span>
          </span>
        </div>
        <button
          data-testid="delete-btn"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="delete"
            src="trash.svg"
          />
        </button>
      </div>
      <button
        className="flex w-full items-center justify-center gap-2 rounded-lg bg-primary p-4 font-bold text-white shadow-lg shadow-primary/50 transition-all duration-300 hover:bg-primary/50"
        data-testid="main-btn"
        type="button"
      >
        <span>
          Checkout
        </span>
      </button>
    </div>
  </div>
</div>
`;
